DT: augment tree and use it to re-order macros

mbolivar opened the pull request:
This reorders the DT macros so they're easier to read.

I've tried to be careful about bisectability by making the output of this PR all the way up to the commit with shortlog `scripts: gen_defines: re-work write_bus() with augmented nodes` match the existing implementation exactly (other than the order in which macros are generated, and some additional comments). This carries forward some behavior I think is questionable, which I've left FIXME comments about.

I've run `sanitycheck --cmake-only -T samples` and wrote a [script to compare the macros](https://gist.github.com/mbolivar/af8256b7b6d2ea448e634075b262f883) before this PR and up to that commit, without any observed differences.

This new way of doing things is much more explicit about which macros are generated in the code, which should hopefully be more readable and make it easier to see things that could be removed.

mbolivar-nordic committed:
scripts: gen_defines: augment edtlib nodes

Add additional attributes to each edtlib.Node we process, before
calling into the write_foo() routines.

This includes the identifier returned by node_ident(), which is used
as the primary identifier for the node, as well as lists for instance
and aliases nodes, and a catchall list that contains all the other
identifiers in addition to the primary one.

Use this information in a new for_each_ident(node, ident) def that
will be put to use in the various write_foo() routines.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_regs() with augmented nodes

Group the macros together by namespace rather than putting all the
BASE_ADDRESS macros together and all the SIZE macros together. E.g.,
all the DT_INST_<x> namespace macros for each node now appear
consecutively.

Add a comment making it clear that this output comes from "regs",
since "BASE_ADDRESS" and "SIZE" are not property names.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
Example output for `/soc/spi@10014000/flash@0` when building hello_world for hifive1_revb:

```
/* BASE_ADDRESS and SIZE macros from the "reg" property */
#define DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0 0x0
#define DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0 0x0
#define DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0
#define DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0
#define DT_INST_0_ISSI_IS25LP128_BASE_ADDRESS_0     DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0
#define DT_INST_0_ISSI_IS25LP128_SIZE_0             DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0
#define DT_INST_0_ISSI_IS25LP128_BASE_ADDRESS       DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0
#define DT_INST_0_ISSI_IS25LP128_SIZE               DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0
#define DT_INST_0_JEDEC_SPI_NOR_BASE_ADDRESS_0      DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0
#define DT_INST_0_JEDEC_SPI_NOR_SIZE_0              DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0
#define DT_INST_0_JEDEC_SPI_NOR_BASE_ADDRESS        DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_BASE_ADDRESS_0
#define DT_INST_0_JEDEC_SPI_NOR_SIZE                DT_SIFIVE_SPI0_10014000_JEDEC_SPI_NOR_0_SIZE_0
```

And for `/soc/interrupt-controller@c000000`:

```
/* BASE_ADDRESS and SIZE macros from the "reg" property */
#define DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_0 0xc000000
#define DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_0         0x2000
#define DT_SIFIVE_PLIC_1_0_0_C000000_PRIO_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_0
#define DT_SIFIVE_PLIC_1_0_0_C000000_PRIO_SIZE      DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_0
#define DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_1 0xc002000
#define DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_1         0x1fe000
#define DT_SIFIVE_PLIC_1_0_0_C000000_IRQ_EN_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_1
#define DT_SIFIVE_PLIC_1_0_0_C000000_IRQ_EN_SIZE    DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_1
#define DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_2 0xc200000
#define DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_2         0x2000000
#define DT_SIFIVE_PLIC_1_0_0_C000000_REG_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_2
#define DT_SIFIVE_PLIC_1_0_0_C000000_REG_SIZE       DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_2
#define DT_INST_0_SIFIVE_PLIC_1_0_0_BASE_ADDRESS_0  DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_0
#define DT_INST_0_SIFIVE_PLIC_1_0_0_SIZE_0          DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_0
#define DT_INST_0_SIFIVE_PLIC_1_0_0_PRIO_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_0
#define DT_INST_0_SIFIVE_PLIC_1_0_0_PRIO_SIZE       DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_0
#define DT_INST_0_SIFIVE_PLIC_1_0_0_BASE_ADDRESS_1  DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_1
#define DT_INST_0_SIFIVE_PLIC_1_0_0_SIZE_1          DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_1
#define DT_INST_0_SIFIVE_PLIC_1_0_0_IRQ_EN_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_1
#define DT_INST_0_SIFIVE_PLIC_1_0_0_IRQ_EN_SIZE     DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_1
#define DT_INST_0_SIFIVE_PLIC_1_0_0_BASE_ADDRESS_2  DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_2
#define DT_INST_0_SIFIVE_PLIC_1_0_0_SIZE_2          DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_2
#define DT_INST_0_SIFIVE_PLIC_1_0_0_REG_BASE_ADDRESS DT_SIFIVE_PLIC_1_0_0_C000000_BASE_ADDRESS_2
#define DT_INST_0_SIFIVE_PLIC_1_0_0_REG_SIZE        DT_SIFIVE_PLIC_1_0_0_C000000_SIZE_2
```

mbolivar commented:
@pabigot -- this also fixes an issue you mentioned in dev-review where you have to special-case code depending on whether there is only one reg or more than one. Now, the BASE_ADDRESS_0 and BASE_ADDRESS defines are both emitted.

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -195,23 +253,69 @@ def relativize(path):
+    out_comment('BASE_ADDRESS and SIZE macros from the "reg" property',
+                blank_before=False)
+    write_regs_for_ident(node.z_legacy_ident)
+    for ident in node.z_inst_idents:
```
Why treat `z_inst_idents` and `z_alias_ident` as unique things?

mbolivar replied:
Just to have a separation between identifiers for instances and identifiers for aliases. I can merge them into one list if that's better.

mbolivar replied:
Hm, I take that back, actually. Notice how write_regs() uses both, but write_irqs() does not use aliases.

mbolivar replied:
That could just be due to a lack of aliases, though. Still working my way through this script converting out_node() to direct calls to out() in this way.

mbolivar-nordic committed:
scripts: gen_defines: re-work write_irqs() with augmented nodes

Use augmented nodes to print macros grouped by namespace.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -105,6 +106,63 @@ def parse_args():
+    #   compatible, plus information from its unit address (or its
+    #   parent's unit address) or its name, and/or its bus. No example
+    #   given since this is a complex and questionably useful form of
+    #   generating an identifier.
```
Wonder if these could be calculated as needed instead, e.g. with a function that takes a node and returns the path identifier. Maybe the `z_` could be dropped then too, if the functions are in `gen_defines.py`.

mbolivar replied:
I realize that efficiency is not necessarily a goal for this script, but one of the benefits of augmenting the tree is that you can reuse the results of `node.edt.compat2enabled[compat].index(node)` by stashing it in z_inst_idents instead of repeating it each time `node_instance_aliases()` is called.

Since your comments so far seem to be about specifics rather than the PR in general, I'll fold in your suggestions and keep working on it. When it's done, `node_instance_aliases()` will be removed entirely.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -195,23 +253,69 @@ def relativize(path):
+        # the primary macros are saved.
+
+        for reg_i, reg in enumerate(node.regs):
+            idx = f"_{reg_i}"
```
I'd remove this variable and pass `f"_{reg_i}"` directly to `write_reg()`.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -195,23 +253,69 @@ def relativize(path):
+
+        for reg_i, reg in enumerate(node.regs):
+            idx = f"_{reg_i}"
+            name = f"{str2ident(reg.name)}_" if reg.name else ""
```
Wonder if this is redundant. `name` is only used within an `if reg.name` below, so it seems it'd never be `""` if it's used.

Maybe `f"{str2ident(reg.name)}_"` could be passed directly to `write_reg()` instead.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -195,23 +253,69 @@ def relativize(path):
+            size = primary_sizes.get(reg_i)
+
+            # Emit BASE_ADDRESS_<INDEX> and SIZE_<INDEX> macros.
+            prim_addr, prim_size = write_reg(ident, reg, addr, size, "", idx)
```
Could move these into the `if addr is None:`, since that's the only place where they're used.

ulfalizer replied:
Never mind. The function needs to be called still.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -195,23 +253,69 @@ def relativize(path):
+            if reg.name:
+                write_reg(ident, reg, addr, size, name, "")
+
+    def write_reg(ident, reg, addr_val, size_val, prefix, suffix):
```
Could get rid of some variables so that you don't have to scan ahead to check if `size_val` and `addr_val` are used elsewhere:

```python
def write_reg(ident, reg, addr, size, prefix, suffix):
    if addr is None:
        addr = hex(reg.addr)

    if size is None:
        size = reg.size

    ...
```

Looks like `addr_val` means "override address from `reg` with this value, if not None." Wonder if that could be handled in some other way. Bit cryptic.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -253,23 +253,69 @@ def relativize(path):
-        if reg.size:
-            write_reg(reg, "SIZE", reg.size)
+    primary_addrs = {}
+    primary_sizes = {}
```
This is a bit confusing to me. Does "primary" mean "path" here too? Noticed that comment above. I'd spell it out in that case.

A nice naming convention is for dicts is `x2y` btw, like `index2addr`.

mbolivar replied:
I'm deliberately being vague about which of the node identifiers is the "primary" identifier. Right now it's the `<COMPAT>_<UNIT_ADDR>` style one. Later on it will be the path-based one. Let me see if I can make this clearer. Thanks.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -253,23 +253,69 @@ def relativize(path):
+            idx = f"_{reg_i}"
+            name = f"{str2ident(reg.name)}_" if reg.name else ""
+            addr = primary_addrs.get(reg_i)
+            size = primary_sizes.get(reg_i)
```
Maybe this could be refactored to remove the `addr = None ... if addr is None:` indirection:

```python
if reg_i not in primary_addr:
    primary_addrs[reg_i] = prim_addr
    primary_sizes[reg_i] = prim_size

reg = primary_addrs[reg_i]
addr = primary_sizes[reg_i]

write_reg(...)
```

Could maybe merge `primary_addrs` and `primary_sizes` too, since they're always accessed together, and maybe use a list, since you never get any "holes" in the indices (though dict is fine if it turns out clearer).

mbolivar-nordic committed:
scripts: gen_defines: re-work write_clocks() with augmented nodes

Mirror the change already done to write_regs().

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_props() with augmented nodes

This is similar to the work already done for regs.

Other than the order in which they appear and comments, the output
before and after this patch should be exactly the same.

We're intentionally leaving some of the helpers in module scope here
to use some of the subroutines elsewhere later on when reworking
write_spi_dev().

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_spi_dev() with augmented nodes

This uses augmented nodes in the same way as done previously.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: re-work write_bus() with augmented nodes

This too is an attempt to reimplement the previous behavior exactly,
modulo the order in which things are defined.

This is the last function which is calling into the previous
implementation's out_node and node_*_alias() functions, so these can
be removed now.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: remove out()'s aliases kwarg

Write the same results explicitly in its only remaining caller.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: add 'labels' attribute to Node class

This corresponds to the attribute by the same name in dtlib.Node.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: tweak out_define() whitespace

Pad node identifiers to 60 characters. This results in better
alignment in practice than the current value of 40, which is a bit
low.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
@pabigot @galak @ulfalizer please revisit.

I've finished the work redoing things in terms of the new augmented tree and made other changes in response to review, then added the new namespaces.

Some follow up work should be done after this to address the FIXME comments mentioned in the PR description.

galak commented:
rebased on master to pull out the doc change that is now merged in master.

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -105,6 +106,103 @@ def parse_args():
+    #
+    # - z_alias_idents: node identifiers based on any /aliases pointing to
+    #   the node in the devicetree source, e.g.:
+    #   ["DT_ALIAS_<NODE's_ALIAS_NAME>"]
```
Missing `["DT_<NODE's_COMPAT>_<NODE's_ALIAS_NAME>"]` form in here (while we want to deprecate it, good to have that in the list)

galak replied:
Also missing docs on `NODEORD`, `PATH`, and `NODELABEL`

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -193,61 +291,389 @@ def relativize(path):
+            # DT_<IDENT>_<PROP_NAME> <PROP_VALUE>
+            ret = {"value": out(name, prop2value(prop))}
+        elif prop.type in ["array", "string-array", "uint8-array"]:
+            # Array types that don't involve phandles are handled by
```
comment needs to clarify `phandle-array` v `phandles`

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -193,61 +291,389 @@ def relativize(path):
+        #
+        # - whether or not to generate an array initializer for "simple"
+        #   (non-phandle) arrays depends on the DTS array type
+        # - arrays involving phandles have special cases depending on
```
Need to clarify `phandles` here.

galak commented:
Add a commit to deal with generation for phandle(s) - based on PR #22770

galak commented:
I guess I'd like to see PATH be considered the primary ident.  When looking for defines or info in the generated file, the NODEORD has not meaning from anything in the DTS so its a bit more painful to search for things.

mbolivar commented:
@galak comments addressed; here's a random example:

```
/*
 * Devicetree node:
 *   /memory@1fff0000
 *
 * Binding (compatible = mmio-sram):
 *   $ZEPHYR_BASE/dts/bindings/sram/mmio-sram.yaml
 *
 * Dependency Ordinal: 4
 *
 * Requires:
 *   0   /
 *
 * Description:
 *   Generic on-chip SRAM description
 */
/* BASE_ADDRESS and SIZE macros from the 'reg' property */
#define DT_PATH_MEMORY_1FFF0000_BASE_ADDRESS                            0x1fff0000
#define DT_PATH_MEMORY_1FFF0000_SIZE                                    65536
#define DT_NODEORD_4_BASE_ADDRESS                                       DT_PATH_MEMORY_1FFF0000_BASE_ADDRESS
#define DT_NODEORD_4_SIZE                                               DT_PATH_MEMORY_1FFF0000_SIZE
#define DT_MMIO_SRAM_1FFF0000_BASE_ADDRESS                              DT_PATH_MEMORY_1FFF0000_BASE_ADDRESS
#define DT_MMIO_SRAM_1FFF0000_SIZE                                      DT_PATH_MEMORY_1FFF0000_SIZE
#define DT_INST_0_MMIO_SRAM_BASE_ADDRESS                                DT_PATH_MEMORY_1FFF0000_BASE_ADDRESS
#define DT_INST_0_MMIO_SRAM_SIZE                                        DT_PATH_MEMORY_1FFF0000_SIZE
#define DT_NODELABEL_SRAM_L_BASE_ADDRESS                                DT_PATH_MEMORY_1FFF0000_BASE_ADDRESS
#define DT_NODELABEL_SRAM_L_SIZE                                        DT_PATH_MEMORY_1FFF0000_SIZE
#define DT_INST_0_MMIO_SRAM                                             1
```

I'll retest the bisectability guarantee when I'm at the office later.

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -105,6 +106,117 @@ def parse_args():
+    # are used to generate macros from it:
+    #
+    # - z_primary_ident: a primary node identifier based on the node's
+    #   path, PATH_<str2ident(node.path>)
```
nit: your formatting here is messed up between < ( > )

mbolivar replied:
Fixed

mbolivar commented:
Removed the ordinal base defines until we have a use case.

mbolivar commented:
pylint output seems stale. trying to kick it.

mbolivar commented:
Pushed and retested; 57dd032fb7 through 61d0ac4f13 ("scripts: gen_defines: tweak out_define() whitespace") are producing the same macros using my script.

galak commented:
@erwango was correct that we need `#define DT_NODELABEL_<NODE-LABEL> 1` to test for existence.

galak reviewed (changes requested):
need to add `#define DT_NODELABEL_<NODE-LABEL> 1` set of defines.

mbolivar commented:
> need to add `#define DT_NODELABEL_<NODE-LABEL> 1` set of defines.

done

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -298,20 +776,29 @@ def write_bus(node):
-
-    # TODO: Handle PWM on STM
-    # TODO: Better document the rules of how we generate things
+    def write_existence_flags_for_ident(node, ident):
```
I wondering if we should make this `DT_EXIST_<....> 1` or something.  As it stands the `#define DT_PATH_SOC_PINMUX_4004A000 1` ends up conflict with the prefix define for phandle(s).

```
/* phandle */
#define DT_PATH_SOC_GPIO_400FF080_NXP_KINETIS_PORT                      DT_PATH_SOC_PINMUX_4004A000
/* path exists */
#define DT_PATH_SOC_PINMUX_4004A000 1
```

mbolivar replied:
Sure, happy to tweak as needed. Is this a blocker for the other work you're doing? I can just restrict it to NODELABEL for now if that's unlikely to get us into trouble.

galak replied:
Its a blocker so lets just restrict to previous defines and NODELABEL.

galak reviewed (changes requested):
Need to tweak the exist defines to either not create on for PATH or change it so its name spaced differently from other PATH defines.

galak committed:
kconfig: Add dt_nodelabel_enabled function

Add a function takes a 'label' and returns "y" if we find an
"enabled" node that has a 'nodelabel' of 'label' in the EDT
otherwise we return "n"

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -193,61 +302,382 @@ def relativize(path):
+    # macros used to identify it, which look like:
+    #
+    # DT_<PRIMARY_NODE_IDENTIFIER>_BASE_ADDRESS_<reg_i>
+    # DT_<PRIMARY_NODE_IDENTIFIER>_SIZE_<reg_i>
```
Could maybe indent code examples by two spaces or the like to make them stand out better from the main text.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -193,61 +302,382 @@ def relativize(path):
+            # DT_<IDENT>_<PROP_NAME> <PROP_VALUE>
+            ret = {"value": out(name, prop2value(prop))}
+        elif prop.type in ["array", "string-array", "uint8-array", "phandles"]:
+            # non-phandle-array array types are handled by
```
*N

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -193,61 +302,382 @@ def relativize(path):
+        out(f"{ident}_{prop_ident}{suffix}", init_macro)
+        if entry_ident:
+            out(f"{ident}_{prop_ident}_{entry_ident}", init_macro)
+
```
Accidental extra blank line.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -278,6 +708,47 @@ def should_write(prop):
+
+    if prop.type == "boolean":
+        return 1 if prop.val else 0
+    elif prop.type == "string":
```
Could put `if` instead of `elif` after the `return`s.

Think `pylint` (and `checkpatch.pl`) warns for it, but disabled it since it's a bit case-to-case.

galak commented:
@mbolivar rebased on master and pulled in a commit to add a kconfig function for use w/nodelabels.

mbolivar commented:
Retested between 2e6983c0f5 and cc9ea0e71c, same macros so bisectability should be OK. Then the remaining parts of the series start changing things.

mbolivar commented:
Addressed @ulfalizer comments

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -298,20 +786,42 @@ def write_bus(node):
+                blank_before=False)
+    for ident in node.z_inst_idents:
+        write_existence_flags_for_ident(node, ident)
+    for ident in node.z_label_idents:
```
Would prefer if we did NOT generate `DT_NODELABEL_FOO 1` and only generated the `DT_EXISTS_NODELABEL_FOO 1`

galak commented:
@mbolivar pushed an update to fix a commit message that mentioned NODEORD.

galak commented:
Might be good to rebase on 22922 and update the docs for DT_PATH, DT_NODELABEL and DT_EXISTS_.

mbolivar commented:
> Might be good to rebase on 22922 and update the docs for DT_PATH, DT_NODELABEL and DT_EXISTS_.

I'm going to rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/22966 instead, once I can remove DNM from that.

mbolivar commented:
Re-worked to remove all the commits which change the generated macros.

mbolivar commented:
Retested macros are the same between e7be2e1bdb and 1be14e9c01.

galak commented:
Merging by hand since the scancode failure is a bug in scancode action.
